ADR 17: Historical header module

cwgoes opened the pull request:
Closes https://github.com/cosmos/cosmos-sdk/issues/5078

- [x] Linked to github-issue with discussion and accepted design OR link to spec that describes this work.
- [x] Updated relevant documentation (`docs/`)
- [x] Re-reviewed `Files changed` in the github PR explorer

______

For Admin Use:
- Added appropriate labels to PR (ex. wip, ready-for-review, docs)
- Reviewers Assigned
- Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

fedekunze commented on docs/architecture/adr-017-historical-header-module.md:
````diff
@@ -0,0 +1,58 @@
+
+```golang
+func (c Context) PreviousHeader(height uint64) tmtypes.Header {
+  // implemented in the context
````
can you complete this function? What if the `height` provided is not in store?

fedekunze commented on docs/architecture/adr-017-historical-header-module.md:
```diff
@@ -0,0 +1,58 @@
+
+## Decision
+
+The application MUST instruct Tendermint to provide requisite information for the application to track the past `n` headers by indicating as such in a field on `abci.ResponseInitChain` containing the value of `n`.
```
I assume this is a concrete  baseapp option passed on the app initialization (i.e `func(*bam.BaseApp)`) ?

cwgoes replied:
I suppose it could be an option on `BaseApp`.

It (the value `n`) could also be read from state, although it can't change once the chain has started.

cwgoes commented on docs/architecture/adr-017-historical-header-module.md:
```diff
@@ -0,0 +1,58 @@
+Tendermint MUST read this field when handling the `abci.ResponseInitChain` response, and then behave as follows:
+- With the first `abci.RequestBeginBlock`, Tendermint MUST send the `n` most recent committed headers in order.
+  - If fewer than `n` previous headers exist, Tendermint MUST send all previous headers.
+- With subsequent `abci.RequestBeginBlock` calls, Tendermint MUST send the the most recent committed header.
```
Only the latest, since the application already has the previously sent headers in memory.

cwgoes commented on docs/architecture/adr-017-historical-header-module.md:
```diff
@@ -0,0 +1,58 @@
+- With subsequent `abci.RequestBeginBlock` calls, Tendermint MUST send the the most recent committed header.
+
+The application MUST use this header data provided by `abci.RequestBeginBlock` to track the past `n` committed headers in the `BaseApp`:
+- When handling the first `abci.RequestBeginBlock` invocation, the application must store all `n` committed headers in memory.
```
In memory (as a field on the baseapp, which is copied into the context, I suppose).

cwgoes commented:
Decided to scrap this in favour of just duplicating state. I will revise.

AdityaSripal commented:
Seems like we're scrapping this idea for a new one? Is there a short description of the new approach anywhere?

Seems pretty trivial to handle this completely on the application side rather than making any changes to ABCI.

Since `n` is an application-parameter and the current BlockHeader and ValidatorSet is introspectable by application using the Context, we can simply save `ctx.BlockHeader` and the current ValidatorSet on each `BeginBlock` under the current height and then prune the store to ensure we always have `n` most recent historical information. So we have some module `history` with a BeginBlocker like:

```go
func BeginBlock(ctx sdk.Context, req abci.RequestBeginBlock) abci.ResponseBeginBlock {
    h := HistoricalInfo{
         Header: ctx.BlockHeader(),
         ValSet:   keeper.StakingKeeper.GetAllValidators(ctx),
    }
    keeper.SetHistoricalInfo(ctx, key=ctx.BlockHeight(), val=encode(h))
    keeper.PruneHistoricalInfo(ctx, key = ctx.BlockHeight() - params.N)
}
```

This can then be queried by the IBC modules. If this is already the new approach we're going for, I'm in favor of it 👍

cwgoes commented:
> Seems like we're scrapping this idea for a new one? Is there a short description of the new approach anywhere?
> 
> Seems pretty trivial to handle this completely on the application side rather than making any changes to ABCI.
> 
> Since `n` is an application-parameter and the current BlockHeader and ValidatorSet is introspectable by application using the Context, we can simply save `ctx.BlockHeader` and the current ValidatorSet on each `BeginBlock` under the current height and then prune the store to ensure we always have `n` most recent historical information. So we have some module `history` with a BeginBlocker like:
> 
> ```go
> func BeginBlock(ctx sdk.Context, req abci.RequestBeginBlock) abci.ResponseBeginBlock {
>     h := HistoricalInfo{
>          Header: ctx.BlockHeader(),
>          ValSet: keeper.StakingKeeper.GetAllValidators(ctx),
>     }
>     keeper.SetHistoricalInfo(ctx, key=ctx.BlockHeight(), val=encode(h))
>     keeper.PruneHistoricalInfo(ctx, key = ctx.BlockHeight() - params.N)
> }
> ```
> 
> This can then be queried by the IBC modules. If this is already the new approach we're going for, I'm in favor of it

Yes; precisely. I might just copy your code!

(I had written the previous version in an attempt to save space, since we will now be storing headers on disk in both Tendermint and the SDK - duplicating data - but after discussion with the Tendermint team we agreed that the complexity wasn't worth it.)

cwgoes commented:
Technically we don't need to store the validator set; we can just store the hash and have the handshake require the validator set be provided as an additional parameter (store abstraction in `rc5` of the IBC spec makes this easier). We can store it if we want though.

fedekunze commented on docs/architecture/adr-017-historical-header-module.md:
````diff
@@ -0,0 +1,59 @@
+The application MUST store this information by storing new headers immediately when handling `abci.RequestBeginBlock`:
+
+```golang
+func BeginBlock(ctx sdk.Context, keeper HistoricalHeaderKeeper, req abci.RequestBeginBlock) abci.ResponseBeginBlock {
````
is this part of a new module?

cwgoes replied:
It can be. If we do anything similar in another module perhaps we could combine them.

cwgoes replied:
Added a comment about this.

fedekunze commented on docs/architecture/adr-017-historical-header-module.md:
```diff
@@ -0,0 +1,59 @@
+func BeginBlock(ctx sdk.Context, keeper HistoricalHeaderKeeper, req abci.RequestBeginBlock) abci.ResponseBeginBlock {
+  info := HistoricalInfo{
+    Header: ctx.BlockHeader(),
+    ValSet: keeper.StakingKeeper.GetAllValidators(ctx),
```
do we need the validator set to be sorted (eg by power)?

cwgoes replied:
I believe it would be sorted already? It just needs to be canonically ordered.

cwgoes replied:
Added a comment.

fedekunze commented on docs/architecture/adr-017-historical-header-module.md:
```diff
@@ -0,0 +1,59 @@
+
+### Negative
+
+- Duplicates header data in Tendermint & the application (additional disk usage)
```
do you see any alternative (medium/long term) to fix this issue?

cwgoes replied:
Yes. https://github.com/tendermint/tendermint/issues/4210

cwgoes replied:
Added a comment linking to this.

fedekunze commented on docs/architecture/adr-017-historical-header-module.md:
```diff
@@ -0,0 +1,59 @@
+
+### Negative
+
+- Duplicates header data in Tendermint & the application (additional disk usage)
```
```suggestion
- Duplicates `n` headers data in Tendermint & the application (additional disk usage)
```

AdityaSripal commented:
Agreed, we can just store the hash if there's a simple way to retrieve the validator set for a given height. How would the relayer get the validator set. Is there already a query for it on the staking store?

cwgoes commented:
> Agreed, we can just store the hash if there's a simple way to retrieve the validator set for a given height. How would the relayer get the validator set. Is there already a query for it on the staking store?

It needs to be queried from Tendermint, I expect, as the SDK can't necessarily calculate it (information is lost if a validator unbonds, rebonds, for example).

AdityaSripal commented:
OK, so long as its queryable by the relayer as-needed rather than forcing the relayer to keep track of the validator sets themselves in their own local db; I am ok with storing hash

cwgoes commented:
I think I've addressed all the comments - not sure what the SDK merge process is these days - @alexanderbez did you want to quickly review (+ merge if satisfactory)?

cwgoes commented:
> ACK.
> 
> 1. Can you add the ADR to the TOC in `docs/architecture/README.MD`?
> 2. Since this doesn't process any txs, is it worth having it as a new module or can we retrofit into `x/staking`? Not a blocker, mainly just a question of best design.

1. Done.
2. Noted that `x/staking` would be a good existing module to use.

fedekunze commented on docs/architecture/adr-017-historical-header-module.md:
```diff
@@ -0,0 +1,61 @@
+
+Alternatively, the application MAY store only the hash of the validator set.
+
+The application MUST make these past `n` committed headers available for querying by SDK modules through the `Keeper`'s `GetHistoricalInfo` function. This MAY be implemented in a new module, or it MAY also be integrated into an existing one (likely `x/staking`).
```
what about storing it on the canonical IBC Keeper? I don't see the need for blockchains that are not using IBC to store the extra headers.

cwgoes replied:
Sure, added `x/ibc` as another suggestion.

cwgoes commented:
(also, why does the simulation fail in some of the above CI runs? this PR only alters `docs`)

alexanderbez commented:
@cwgoes most likely because each successive commit kills any previous CI pipeline (so we only have 1 CI pipeline running per commit/per PR).
